Backport of some of MSE related changes from wpe-2.22

emutavchi opened the pull request:
Original PR for wpe-2.22: https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/697/

emutavchi committed:
[MSE] Fix occasional video playback stalls after seek

emutavchi committed:
[Westeros-sink] Initial playbin3 support

emutavchi committed:
Disable endcoded data replacement for images with async decoding

Replacing data during async decoding causes random crashes

emutavchi committed:
Fix loss of ReadyForMoreSamples notifications causing video freeze

When delegating the notification of "ReadyForMoreSamples" to another thread via
the "MainThreadNotifier" class, if a prior notification request is still
pending, the new request is dropped. As notification requests can occur
concurrently when multiple streams are present (e.g. audio and video), a
notification may get lost, because the notify only tracks pending
notifications per notification type, and does not have knowledge of streams.
I.e. if the notifier has a pending "ReadyForMoreSamples" notification
for stream 1 and we are requesting a notification for stream 2, then the
notification for stream 2 is ignored/dropped.

While this kind of notification may be triggered again, there are corner
cases when such re-trigger is not happening if it was already requested
(e.g. empty queue). To ensure notifications are triggered for each
stream, we use stream type specific notifications.

An observed effect of this issue is video freezing with audio continuing

eocanha commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -280,7 +280,8 @@ void MediaSource::completeSeek()
     // 4. Resume the seek algorithm at the "Await a stable state" step.
     m_private->seekCompleted();
 
-    monitorSourceBuffers();
```
The current monitorSourceBuffers() implementation already checks m_pendingSeekTime.isInvalid() before calling completeSeek() so, in theory, the recursion should already be stopped there and this patch shouldn't be needed.

Have you been able to reproduce the infinite recursion on 2.28? Do you have a backtrace to check?

emutavchi replied:
Indeed. I missed that. Let me remove the commit from the PR.

emutavchi replied:
@eocanha I've removed commit from PR.

eocanha commented on Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
```diff
@@ -798,7 +798,7 @@ static PlatformTimeRanges removeSamplesFromTrackBuffer(const DecodeOrderSampleMa
         }
 
-        auto endIterator = trackBuffer.samples.presentationOrder().findSampleStartingOnOrAfterPresentationTime(erasedEnd);
+        auto endIterator = trackBuffer.samples.presentationOrder().findSampleContainingOrAfterPresentationTime(erasedEnd);
```
Can you explain a practical example where findSampleStartingOnOrAfterPresentationTime() causes problems and findSampleContainingOrAfterPresentationTime() solves them? Having a good practical case for justifying this change would be helpful in order to submit this change to WebKit upstream (our policy now for wpe-2.28 commits).

emutavchi replied:
This change fixes high memory usage in apps that do TS -> MP4 transmuxing. The high memory usage is observed in some streams that have a slight overlap in media samples (usually happens on video quality change), when application applies its own logic for samples removal (that looks like sourceBuffer.remove(sourceBuffer.buffered.start(0), ...)). 

Current implementation reports a buffered range that is inconsistent with the actual samples that are still in the buffer (after range removal). The reported range excludes one sample, allowing that sample to "leak". Such a sample holds a reference to the entire appended chunk (due to in-place demuxing in GStreamer MSE backend), but it accounts only for a small size in the buffer. So eviction algorithm doesn't kick in.

Here you can find a mock test that simulates the problem: https://github.com/emutavchi/tests/blob/master/mse/media-source-remove.html. You'll need WebKit with Mock MSE and enable MSE debug logs to see the problem. Here is the output observed without change (range removal part):

```
[WPEWebKit:MediaSource:-] SourceBuffer::remove(29ACFEDF) start = {"value":0}, end = {"value":3.96}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) start = {"value":0}, end = {"value":3.96}          
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) removing sample {"pts":{"value":0,"numerator":0,"denominator":1000,"flags":1},"dts":{"value":0,"numerator":0,"denominator":1000,"flags":1},"duration":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"flags":1,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) removing sample {"pts":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"dts":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"duration":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) removing sample {"pts":{"value":2,"numerator":2000,"denominator":1000,"flags":1},"dts":{"value":2,"numerator":2000,"denominator":1000,"flags":1},"duration":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) removing sample {"pts":{"value":3,"numerator":3000,"denominator":1000,"flags":1},"dts":{"value":3,"numerator":3000,"denominator":1000,"flags":1},"duration":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) removed 256, start = {"value":0,"numerator":0,"denominator":1000,"flags":1}, end = {"value":4,"numerator":4000,"denominator":1000,"flags":1}                                                                                                

[WPEWebKit:MediaSource:-] SourceBuffer::remove(29ACFEDF) start = {"value":4.29}, end = {"value":5}
[WPEWebKit:Media:-] HTMLMediaElement::scheduleUpdatePlayState(29ACFEDF) lambda(), task fired
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) start = {"value":4.29}, end = {"value":5}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) removing sample {"pts":{"value":4.29,"numerator":4290,"denominator":1000,"flags":1},"dts":{"value":4.29,"numerator":4290,"denominator":1000,"flags":1},"duration":{"value":0.33,"numerator":330,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) removing sample {"pts":{"value":4.62,"numerator":4620,"denominator":1000,"flags":1},"dts":{"value":4.62,"numerator":4620,"denominator":1000,"flags":1},"duration":{"value":0.33,"numerator":330,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) removing sample {"pts":{"value":4.95,"numerator":4950,"denominator":1000,"flags":1},"dts":{"value":4.95,"numerator":4950,"denominator":1000,"flags":1},"duration":{"value":0.33,"numerator":330,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(29ACFEDF) removed 192, start = {"value":4.29,"numerator":4290,"denominator":1000,"flags":1}, end = {"value":5.28,"numerator":5280,"denominator":1000,"flags":1} 
```

note that sample at PTS 3.96 "leaked".

here is the output after the fix:
```
[WPEWebKit:MediaSource:-] SourceBuffer::remove(1FCB8181) start = {"value":0}, end = {"value":3.96}                                                                                                         
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) start = {"value":0}, end = {"value":3.96}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) removing sample {"pts":{"value":0,"numerator":0,"denominator":1000,"flags":1},"dts":{"value":0,"numerator":0,"denominator":1000,"flags"
:1},"duration":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"flags":1,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) removing sample {"pts":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"dts":{"value":1,"numerator":1000,"denominator":1000,"
flags":1},"duration":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) removing sample {"pts":{"value":2,"numerator":2000,"denominator":1000,"flags":1},"dts":{"value":2,"numerator":2000,"denominator":1000,"
flags":1},"duration":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) removing sample {"pts":{"value":3,"numerator":3000,"denominator":1000,"flags":1},"dts":{"value":3,"numerator":3000,"denominator":1000,"
flags":1},"duration":{"value":1,"numerator":1000,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) removed 256, start = {"value":0,"numerator":0,"denominator":1000,"flags":1}, end = {"value":4,"numerator":4000,"denominator":1000,"flag
s":1}

[WPEWebKit:MediaSource:-] SourceBuffer::remove(1FCB8181) start = {"value":4}, end = {"value":5}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) start = {"value":4}, end = {"value":5}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) removing sample {"pts":{"value":3.96,"numerator":3960,"denominator":1000,"flags":1},"dts":{"value":3.96,"numerator":3960,"denominator":
1000,"flags":1},"duration":{"value":0.33,"numerator":330,"denominator":1000,"flags":1},"flags":1,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) removing sample {"pts":{"value":4.29,"numerator":4290,"denominator":1000,"flags":1},"dts":{"value":4.29,"numerator":4290,"denominator":
1000,"flags":1},"duration":{"value":0.33,"numerator":330,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) removing sample {"pts":{"value":4.62,"numerator":4620,"denominator":1000,"flags":1},"dts":{"value":4.62,"numerator":4620,"denominator":
1000,"flags":1},"duration":{"value":0.33,"numerator":330,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) removing sample {"pts":{"value":4.95,"numerator":4950,"denominator":1000,"flags":1},"dts":{"value":4.95,"numerator":4950,"denominator":
1000,"flags":1},"duration":{"value":0.33,"numerator":330,"denominator":1000,"flags":1},"flags":0,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::removeCodedFrames(1FCB8181) removed 256, start = {"value":3.96,"numerator":3960,"denominator":1000,"flags":1}, end = {"value":5.28,"numerator":5280,"denominator":1
000,"flags":1}
```

jernoble replied:
This result (sample at PTS 3.96 "leaked") is correct:

<https://www.w3.org/TR/media-source/#sourcebuffer-coded-frame-removal>

> Remove all media data, from this [track buffer](https://www.w3.org/TR/media-source/#track-buffer), that contain starting timestamps greater than or equal to start and less than the remove end timestamp.

Calling `sourceBuffer.remove(sourceBuffer.buffered.start(0), 3.96)` should not remove a sample whose presentation timestamp is `3.96` (and a sync sample) because that sample is not "less than the remove end timestamp".

emutavchi replied:
Hi @jernoble, thanks for looking into this. The problem is with the buffered range report. The fix doesn't change removal behavior, it only affects the buffered range report. Currently, as you can see, the buffered range starts from 4.29 (after the first removal). This is not correct, 3.96 sample is still there in the buffer.

jernoble replied:
Ah I see! Let me look more into why your proposed change has the effect you're trying to achieve and respond in a bit.

jernoble replied:
So there's another problem with this content, in that the sample at `pts=3, duration=1` overlaps with the sample at `pts=3.96, duration=.33`. And that's causing a problem with calculating the end time of the removed range, which is what in turn causes this behavior.

I don't think that changing the method from `startingOnOrAfter` to `containingOnOrAfter` is actually the correct fix for this issue. The method was written with the assumption that samples in a given track do not overlap. After all, appending an overlapping sample _should_ evict the sample it overlaps with. I'm guessing there's a "fudge factor" somewhere that's preventing this eviction from happening.

A more correct fix would be to address why the samples are overlapping in the first place. If a fudge factor is causing a sample that would otherwise be evicted to not be, that sample should have its duration truncated (in the case of a video sample) to avoid the overlap. However, this is a much larger change, impacting much more than the bug you're trying to fix.

If that fix is not possible, the algorithm itself should be re-authored to take overlapping samples into account. For example, the loop immediately above the proposed change could only create a "erased range" spanning from the PTS of one sample to the PTS of the next, then expanding the erased range to the next (un-removed) sample's PTS.

Alternatively, if it is the case that the only way an overlapping sample is allowed to exist in the TrackBuffer is due to a fudge factor, another more correct fix would be to leave `findSampleStartingOnOrAfterPresentationTime()`, but decrement `erasedEnd` by `fudgeFactor`, to ensure you find the remaining sample.

jernoble replied:
Whichever change you decide to go with, please feel free to post it on bugs.webkit.com and CC me; I'd be happy to work with you in getting a fix into the tree.

emutavchi replied:
> So there's another problem with this content, in that the sample at pts=3, duration=1 overlaps with the sample at pts=3.96, duration=.33. And that's causing a problem with calculating the end time of the removed range, which is what in turn causes this behavior.

Right, this is what we observe on some live streams in various apps that do TS -> MP4 transmuxing. 

> I don't think that changing the method from startingOnOrAfter to containingOnOrAfter is actually the correct fix for this issue.

The end of the erased range would point either into an empty space or overlapped sample. if it points into an overlapped sample, I think it makes sense to not extend the erased range that would include that sample duration 

> Alternatively, if it is the case that the only way an overlapping sample is allowed to exist in the TrackBuffer is due to a fudge factor

I don't think sample replacement logic should kick in in this case (if I interpreted the MSE spec correctly).

> Whichever change you decide to go with, please feel free to post it on bugs.webkit.com and CC me; I'd be happy to work with you in getting a fix into the tree.

I still think that solution with containingOnOrAfter is good enough to fix the increased memory usage problem that we observe, without having to do a major rework.

jernoble replied:
> > Alternatively, if it is the case that the only way an overlapping sample is allowed to exist in the TrackBuffer is due to a fudge factor
> 
> I don't think sample replacement logic should kick in in this case (if I interpreted the MSE spec correctly).

Except it does. Samples are not allowed to overlap with previously appended samples, but for various reasons (such as bad muxing or variant switching with different timebases) most UAs allow some amount of overlap. If we were strictly adhering to the spec, the sample at PTS=3.96 would evict the sample at PTS=3. 

The code to reconcile removed samples with buffered ranges looks at the end PTS time of those samples in order to define the removed range; if that removed sample (PTS=3) had its duration truncated to 3.96, the reconciliation algorithm would succeed normally.

> > Whichever change you decide to go with, please feel free to post it on bugs.webkit.com and CC me; I'd be happy to work with you in getting a fix into the tree.
> 
> I still think that solution with containingOnOrAfter is good enough to fix the increased memory usage problem that we observe, without having to do a major rework.

The problem with that solution is it does not take into account the amount of overlap. The solution proposed here would allow for situations that absolutely cannot exist, such a removed sample whose timeline is entirely within a non-removed sample, breaking not only the assumption that samples do not overlap, but that samples monotonically increase in both start PTS and end PTS. I would rather not add the concept if a more limited solution (or, IMO, a more correct one) exists.

emutavchi replied:
> Except it does.

I must be missing somthing. so, I'm trying to understand how this can happen. Please bear with me.
I see 2 places in the spec that are relevant and could trigger removal of buffered samples:

\# 1
<pre>13. If last decode timestamp for track buffer is unset and presentation timestamp falls within the presentation interval of a coded frame in track buffer, then run the following steps: ... </pre>

The last decoded timestamp is set so we can ignore this case.

\# 2
<pre>14. Remove existing coded frames in track buffer:
If highest end timestamp for track buffer is not set:
    Remove all coded frames from track buffer that have a presentation timestamp greater than or equal to presentation timestamp and less than frame end timestamp.
If highest end timestamp for track buffer is set and less than or equal to presentation timestamp:
   Remove all coded frames from track buffer that have a presentation timestamp greater than or equal to highest end timestamp and less than frame end timestamp.
</pre>

the highest end timestamp is set to 4.0 when an overlapped sample is received. it is higher than pts of "arriving" sample (3.96). so this also doesn't trigger replacement.

I've added additionall logging to trace the processing of coded frames (https://github.com/emutavchi/WebKitForWayland/commit/f11ba3f2fb3124cf9bee3f2c53ceccabc7854e1a).

Here you can find full output: https://gist.github.com/emutavchi/47b4e488156a2018fecb4e533eef2603. 

Processing of sample at 3.96 PTS doesn't trigger removal:
```
[WPEWebKit:MediaSource:-] SourceBuffer::sourceBufferPrivateDidReceiveSample(ADC2765D) {"pts":{"value":3.96,"numerator":3960,"denominator":1000,"flags":1},"dts":{"value":3.96,"numerator":3960,"denominator":1000,"flags":1},"duration":{"value":0.33,"numerator":330,"denominator":1000,"flags":1},"flags":1,"presentationSize":{"width":0,"height":0}}
[WPEWebKit:MediaSource:-] SourceBuffer::sourceBufferPrivateDidReceiveSample(ADC2765D) Last decode time stamp is valid, carry on
[WPEWebKit:MediaSource:-] SourceBuffer::sourceBufferPrivateDidReceiveSample(ADC2765D) HighestPresentation timestamp is valid, {"value":4,"numerator":4000,"denominator":1000,"flags":1}
[WPEWebKit:MediaSource:-] SourceBuffer::sourceBufferPrivateDidReceiveSample(ADC2765D) HighestPresentation timestamp is valid, no overlap found!
[WPEWebKit:MediaSource:-] SourceBuffer::sourceBufferPrivateDidReceiveSample(ADC2765D) erasedSamples is empty
```

so, unless there is something else in the spec that would trigger replacement (and not implementet yet), I conclude that overlapping of media samples is allowed.

jernoble replied:
> > Except it does.
> 
> I must be missing somthing. so, I'm trying to understand how this can happen. Please bear with me. I see 2 places in the spec that are relevant and could trigger removal of buffered samples:
> 
> # 1
> 
> 13. If last decode timestamp for track buffer is unset and presentation timestamp falls within the presentation interval of a coded frame in track buffer, then run the following steps: ... 
> The last decoded timestamp is set so we can ignore this case.

If a container is generating overlapping samples within a coded frame group, it can only be the result of either a bad mux or a demuxing error. The spec has an explicit requirement for monotonically increasing decode timestamps, and an implicit requirement for the sample end times to be well ordered. I'd have to check the ISO BMFF spec, but it should not be the case that samples within a container can be overlapping.

Your test case doesn't reset the (mock) parser between sample `PTS=3` and `PTS=3.96`. Does the in-the-wild behavior you're trying to fix reset the parser (with a new initialization segment) between appends?

jernoble replied:
(As for the Mock test case output and logging, you could do all kind of weird things with the Mock "demuxer" and get strange behavior; including having non-monotonically increasing decode times, completely overlapping samples, etc. The Mock is just a tool to allow for cross-platform testing, that's all.)

emutavchi replied:
I agree that the mock test is oversimplified. I'm sorry if that led to confusion. App does append init segment on quality change right before appending the overlapping samples.  

Here is an example that splits append into two and I think it'd better show the problem: https://raw.githubusercontent.com/emutavchi/tests/master/mse/media-source-remove2.html. I've also included a parser reset (abort) call before appending the second init segment (I'm not sure if the app did that, and I don't have logs from the reproduction case at the moment to confirm).

The behavior didn't change though, there was no sample replaced: https://gist.github.com/emutavchi/4820b0abd59f9e7645663b018d4b826f

Note that PTS and DTS are monotonically increasing.

emutavchi replied:
And yes, I agree that this could be an issue with transmuxer on the web app side. but, we cannot fix that.

jernoble replied:
I'd like to recognize that you are indeed correct; the spec does allow for overlapping video samples. I had misremembered; the behavior for audio and text samples is different than for video samples. Personally, I think that's a specification bug, and allowing samples to overlap in the TrackBuffer is confusing and will lead to errors like this one. I would support an addition to MSE that added a "video frame splicing" step similar to the audio and text splice frame algorithm which would simply truncate the duration of the previous sample.  But as the spec stands, that is not required.

That said, I don't think this affects what approach would be the most correct solution to the problem you are trying to solve. Given that overlapping frames can exist in the TrackBuffer, the algorithm's use of `erasedEnd` will inherently cause problems. I still think a more correct solution would be to re-write that section of the algorithm to search the TrackBuffer  using `findSampleStartingAfterPresentationTime()` and `erasedStart`.  Given that samples can overlap, using `findSampleContainingOrAfterPresentationTime()` with `erasedEnd` can still give incorrect results if multiple samples have presentation ranges containing that time.  Because that method uses a binary search algorithm, which sample is found in that scenario is non-deterministic. 

Alternatively, rather than using the sample's duration to calculate the `endTime` of the erased range, the algorithm could use the next remaining sample in PTS order to create the erased range. This would tackle the problem at the correct place, rather than adjusting the erased range to account for overlapping later in the algorithm.

emutavchi replied:
I've updated as suggested to use findSampleStartingAfterPresentationTime() and erasedStart. https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/797/commits/880104d32c33cf39f12e42920a4f2bf25df2fd65

eocanha replied:
I've ported the patch and the layout test upstream as https://bugs.webkit.org/show_bug.cgi?id=237780, just in case @jernoble wants to review it there.

eocanha commented on Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.h:
```diff
@@ -85,6 +85,8 @@ class MediaPlayerPrivateGStreamerMSE : public MediaPlayerPrivateGStreamer {
     WTFLogChannel& logChannel() const final { return WebCore::LogMediaSource; }
 #endif
 
+   bool gstSeekCompleted() const { return m_gstSeekCompleted; }
```
Reminder to myself: This commit isn't upstreamable.

eocanha commented on Source/WebCore/platform/graphics/holepunch/MediaPlayerPrivateHolePunch.h:
```diff
@@ -93,20 +93,24 @@ class MediaPlayerPrivateHolePunch : public MediaPlayerPrivateInterface, public C
 #endif
 
+protected:
+    MediaPlayer* m_player;
```
Why is this change to "protected" needed? Is there going to be a MediaPlayerPrivateHolePunch subclass somewhere?

emutavchi replied:
This was a straightforward cherry-pick. I don't remember why it was needed in 2.22 branch, but it looks like I can remove it in 2.28. I'll update the commit,

eocanha commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -2916,6 +2916,7 @@ void MediaPlayerPrivateGStreamer::createGSTPlayBin(const URL& url, const String&
 #endif
     }), this);
 
+    g_signal_connect_swapped(m_pipeline.get(), "element-setup", G_CALLBACK(elementSetupCallback), this);
```
Reminder to myself: I don't think this commit is upstreamable, as there's no support for Westeros sink upstream.

eocanha commented on Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:
```diff
@@ -134,10 +134,22 @@ static void enabledAppsrcNeedData(GstAppSrc* appsrc, guint, gpointer userData)
 
-        if (appsrcStream && appsrcStream->type != WebCore::Invalid)
-            webKitMediaSrc->priv->notifier->notify(WebKitMediaSrcMainThreadNotification::ReadyForMoreSamples, [webKitMediaSrc, appsrcStream] {
+        if (appsrcStream && appsrcStream->type != WebCore::Invalid) {
```
This commit is useful, but is going to be superseeded by a patch @calvaris is working on, so let's better not merge it into wpe-2.28 yet.
See the original discussion about the topic in https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/791

emutavchi replied:
@eocanha @calvaris are the changes you're working on will land any time soon? if not, I'd request this change to be merged.

calvaris replied:
I don't mind merging this now. I'll fix what I need to fix later.

eocanha reviewed (commented):
Instead of directly merging this PR into wpe-2.28, I'm going to try to upstream the relevant commits and then backport them to wpe-2.28. I'll manually commit those commits that aren't upstreamable.
I've also left some questions for which it would be nice to get an answer. They would be helpful to make a case for upstreaming the related changes.

calvaris commented:
I might be interesting to file this kind of PRs in smallest chunks in the future.

emutavchi committed:
[GStreamer] Handle gst pad error as format error
https://bugs.webkit.org/show_bug.cgi?id=237526

Reviewed by Xabier Rodriguez-Calvar.

This patch is authored by Eugene Mutavchi <Ievgen_Mutavchi@comcast.com>
See: https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/797

* platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:

git-svn-id: http://svn.webkit.org/repository/webkit/trunk@290984 268f45cc-cd09-0410-ab3c-d52691b4dbfc

eocanha commented on Source/WebCore/platform/graphics/MediaPlayer.cpp:
```diff
@@ -1229,7 +1229,7 @@ void MediaPlayer::networkStateChanged()
     if (m_private->networkState() >= MediaPlayer::NetworkState::FormatError && m_private->readyState() < MediaPlayer::ReadyState::HaveMetadata) {
         client().mediaPlayerEngineFailedToLoad();
-        if (!m_activeEngineIdentifier && installedMediaEngines().size() > 1 && (m_contentType.isEmpty() || nextBestMediaEngine(m_currentMediaEngine))) {
+        if (!m_activeEngineIdentifier && installedMediaEngines().size() > 1 && !m_contentType.isEmpty() && nextBestMediaEngine(m_currentMediaEngine)) {
```
The original behaviour of loading with next media engine 
 in reloadTimerFired() when contentType is empty has been there since https://bugs.webkit.org/show_bug.cgi?id=58462, more than 10 years ago.
 Do you remember why this change was needed?

emutavchi replied:
This is needed to notify html media element about the error, and preserve the player with an actual error message. otherwise, it just silently consumes the error while trying to load next engine.

eocanha replied:
This caused regressions in Apple ports upstream, so I reverted the line to avoid player private regeneration and added code to cache the player private error in MediaPlayer as m_lastErrorMessage. That way, the error code of the disposed player private can be accessed after regeneration and the regressions are avoided.

emutavchi committed:
[MSE][SourceBuffer] Fix creating float PTS/DTS when dividing sample
https://bugs.webkit.org/show_bug.cgi?id=237528

Reviewed by Xabier Rodriguez-Calvar.

The timestamps used for sample division in SourceBuffer aren't being
aligned with the same rounding margin used in
SourceBuffer::sourceBufferPrivateDidReceiveSample().

This patch is authored by Eugene Mutavchi <Ievgen_Mutavchi@comcast.com>
See: https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/797

* platform/graphics/SourceBufferPrivate.cpp: Extracted roundTowardsTimeScaleWithRoundingMargin lambda as a function and used it from removeCodedFrames(), on top of the current usage from sourceBufferPrivateDidReceiveSample().

git-svn-id: http://svn.webkit.org/repository/webkit/trunk@291101 268f45cc-cd09-0410-ab3c-d52691b4dbfc

emutavchi committed:
[MSE][SourceBuffer] Fix removal of paddings from buffered range
https://bugs.webkit.org/show_bug.cgi?id=237780

Reviewed by Jer Noble.

Source/WebCore:

When removing a range ending up in a sample S2 which overlaps slightly with a
previous one S1 but still extends beyond the removal range, the buffered range
computing algorithm can become confused removing the range paddings and
completely skip the S2 sample (instead of accounting it as part of the still
buffered ranges). The sample S2 is leaked and becomes unnoticed in the buffered
ranges.

This patch changes the way in which buffered ranges boundaries are computed in
such cases.

This patch is authored by Eugene Mutavchi <Ievgen_Mutavchi@comcast.com>
See: https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/797#discussion_r818973424

Test: media/media-source/media-source-remove-overlapped.html

* platform/graphics/SourceBufferPrivate.cpp: End the additionalErasedRanges iterator in removeSamplesFromTrackBuffer() in a sample starting after erasedStart, instead of starting on or after erasedEnd.

LayoutTests:

This tests that overlapped samples in the boundaries of removal ranges aren't leaked.

This patch is authored by Eugene Mutavchi <Ievgen_Mutavchi@comcast.com>
See: https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/797

* media/media-source/media-source-remove-overlapped-expected.txt: Added.
* media/media-source/media-source-remove-overlapped.html: Added.

git-svn-id: http://svn.webkit.org/repository/webkit/trunk@291225 268f45cc-cd09-0410-ab3c-d52691b4dbfc

eocanha commented on Source/WebCore/loader/cache/CachedImage.cpp:
```diff
@@ -711,4 +711,11 @@ CachedResource::RevalidationDecision CachedImage::makeRevalidationDecision(Cache
     return CachedResource::makeRevalidationDecision(cachePolicy);
 }
 
+bool CachedImage::mayTryReplaceEncodedData() const {
```
This commit is related to Graphics, away from my area of expertise (multimedia). Maybe @magomez might consider to upstream it. Otherwise I'm keeping it downstream only.

magomez replied:
Async decoding was a relatively new feature in 2.22, but evolved a lot since then upstream, and this is probably not a problem anymore there. This should be kept as a downstream fix.

emutavchi committed:
[GStreamer] clarify playback errors with MediaError.message
https://bugs.webkit.org/show_bug.cgi?id=237602

Reviewed by Philippe Normand.

This patch is co-authored by Eugene Mutavchi <Ievgen_Mutavchi@comcast.com>
See: https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/797

* html/HTMLMediaElement.cpp: Use internal last error message from the player, if available.
* platform/graphics/MediaPlayer.cpp: Cache the last error message from the player private and update it when NetworkState changes to an error state. This allows the player private to be regenerated (as it has always been) while providing access to the former player private error after that.
* platform/graphics/MediaPlayer.h: Added lastErrorMessage() method to get the last internal error message from the player private.
* platform/graphics/MediaPlayerPrivate.h: Default lastErrorMessage() empty implementation.
* platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp: Detect decryption and no-key errors.
* platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h: Added errorMessage() to expose the internal error message.
* platform/graphics/gstreamer/eme/WebKitCommonEncryptionDecryptorGStreamer.cpp: Post multiple kind of errors to the GstBus to notify upper layers instead of just logging them.

git-svn-id: http://svn.webkit.org/repository/webkit/trunk@291290 268f45cc-cd09-0410-ab3c-d52691b4dbfc

emutavchi committed:
[GStreamer] Avoid auto-selection of hole punching player
https://bugs.webkit.org/show_bug.cgi?id=237828

Reviewed by Xabier Rodriguez-Calvar.

The hole punch player private shouldn't be automatically selected when
the mime type is empty or unsupported by that specific player
("video/holepunch" in practice). This patch returns a FormatError
networkState in those cases, forcing the selection algorithm to
discard the hole punch player private.

This patch is authored by Eugene Mutavchi <Ievgen_Mutavchi@comcast.com>
See: https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/797

* platform/graphics/holepunch/MediaPlayerPrivateHolePunch.cpp: Implement networkState setter. Set FormatError networkState on load when the mime type is empty or unknown.
* platform/graphics/holepunch/MediaPlayerPrivateHolePunch.h: Refactor networkState management. Move load() implementation to cpp file.

git-svn-id: http://svn.webkit.org/repository/webkit/trunk@291337 268f45cc-cd09-0410-ab3c-d52691b4dbfc
